src/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -99,19 +102,58 @@ public string GetHeaderValue(ReadOnlySpan<byte> headerValue)
             }
 
             // If it's a known header value, use the known value instead of allocating a new string.
-            if (_knownHeader != null && _knownHeader.KnownValues != null)
+            if (_knownHeader != null)
             {
-                string[] knownValues = _knownHeader.KnownValues;
-                for (int i = 0; i < knownValues.Length; i++)
+                if (_knownHeader.KnownValues != null)
+                {
+                    string[] knownValues = _knownHeader.KnownValues;
+                    for (int i = 0; i < knownValues.Length; i++)
+                    {
+                        if (ByteArrayHelpers.EqualsOrdinalAsciiIgnoreCase(knownValues[i], headerValue))
+                        {
+                            return knownValues[i];
+                        }
+                    }
+                }
+
+                if (KnownHeader == KnownHeaders.Location)
                 {
-                    if (ByteArrayHelpers.EqualsOrdinalAsciiIgnoreCase(knownValues[i], headerValue))
+                    // Normally Location should be in ISO-8859-1, ocassionally some servers respond with UTF-8 though
+                    if (TryDecodeUtf8(headerValue, out string decoded))
                     {
-                        return knownValues[i];
+                        return decoded;
                     }
                 }
             }
 
             return HttpRuleParser.DefaultHttpEncoding.GetString(headerValue);
         }
+
+        private static bool TryDecodeUtf8(ReadOnlySpan<byte> input, out string decoded)
+        {
+            // TODO: Utilize vectorization helpers here if/when they're made public
+            bool possibleUtf8 = false;
+            for (int i = 0; i < input.Length; i++)
+            {
+                if (input[i] > 127)
```

Comment:
Is this worth doing? If so why?

If the header has no non-ASCII values, then it will decode the same way whether we try to decode as UTF-8 or ISO-8859-1. So why not just try to decode as UTF-8, and fall back to ISO-8859-1 only if this fails?

Reply:
I think this might be true so I'll investigate this more. Currently I have followed what desktop was doing.

Reply:
yes, seems this should be equivalent